server/portal/apps/projects/workspace_operations/shared_workspace_operations.py

```diff
@@ -0,0 +1,363 @@
+    data = {"key_value": publicKey, "tags": [{"name": "system", "value": system_id}]}
+    response = requests.post(url, json=data, headers=headers)
+    response.raise_for_status
+    return response.status_code
+
+
+def set_workspace_permissions(client: Tapis, username: str, system_id: str, role: str):
+    """Apply read/write/execute permissions to a user on a system."""
+
+    system_pems = {
+         "reader": ["READ", "EXECUTE"],
+         "writer": ["READ", "EXECUTE"]
+    }
+
+    files_pems = {
+         "reader": "READ",
+         "writer": "MODIFY"
+    }
+
+    logger.info(f"Adding {username} permissions to Tapis system {system_id}")
+    client.systems.grantUserPerms(
+        systemId=system_id,
+        userName=username,
+        permissions=system_pems[role])
+
+    if role == "reader":
+        client.systems.revokeUserPerms(systemId=system_id,
+                                       userName=username,
+                                       permissions=["MODIFY"])
+        client.files.deletePermissions(systemId=system_id,
+                                       path="/",
+                                       username=username)
+
+    client.files.grantPermissions(
+        systemId=system_id,
+        path="/",
+        username=username,
+        permission=files_pems[role]
+    )
+
+
+def set_workspace_acls(client, system_id, path, username, operation, role):
+    jwt = client.access_token.access_token
+    url = f"{settings.TAPIS_TENANT_BASEURL}/v3/files/utils/linux/facl/{system_id}/{path}"
+    headers = {"x-tapis-token": jwt}
+
+    operation_map = {
+         "add": "ADD",
+         "remove": "REMOVE"
+    }
+
+    acl_string_map = {
+         "reader": f"d:u:{username}:rX,u:{username}:rX",
+         "writer": f"d:u:{username}:rwX,u:{username}:rwX",
+         "none": f"d:u:{username},u:{username}"
+    }
+
+    data = {"operation": operation_map[operation],
+            "aclString": acl_string_map[role],
+            "recursionMethod": "PHYSICAL"}
+
+    response = requests.post(url, json=data, headers=headers)
+    response.raise_for_status()
+    return response
+
+
+def create_workspace_dir(workspace_id: str) -> str:
+    client = service_account()
+    system_id = settings.PORTAL_PROJECTS_ROOT_SYSTEM_NAME
+    path = f"{workspace_id}"
+    client.files.mkdir(systemId=system_id, path=path)
+    return path
+
+
+def create_workspace_system(client, workspace_id: str, title: str) -> str:
+    system_id = f"{settings.PORTAL_PROJECTS_SYSTEM_PREFIX}.{workspace_id}"
+    system_args = {
+        "id": system_id,
+        "host": settings.PORTAL_PROJECTS_ROOT_HOST,
+        "port": int(settings.PORTAL_PROJECTS_SYSTEM_PORT),
+        "systemType": "LINUX",
+        "defaultAuthnMethod": "PKI_KEYS",
+        "canExec": False,
+        "rootDir": f"{settings.PORTAL_PROJECTS_ROOT_DIR}/{workspace_id}",
+        "notes": {"title": title}
+    }
+    client.systems.createSystem(**system_args)
```

Comment:
I don't think we need to. ID collisions were a problem in the last iteration of CEP because each portal deployment/dev environment maintained its own counter. In this implementation, the counter is part of the root system definition so there's a single source of truth. Collisions are only an issue if 2 people try to create systems at the exact same time, if we want to prevent that we're talking about like... sticking Tapis calls inside atomic transactions or something.

Reply:
Ah right, perfect. In that case, can we also axe the projects models? Or add a task to clean all that unused stuff up?

Reply:
Yeah I'll add a task. We do need to keep the models around until every portal is transitioned, because they're needed to migrate existing projects.

Reply:
Gotcha. Are we thinking a migration script or the like?

Reply:
Yeah Either a command-line util or just a method that can be run from the django shell.

Reply:
👍 Did you want to handle that in this PR or a follow on PR? If the latter, can you make a task?

Reply:
Already a task: https://jira.tacc.utexas.edu/browse/TV3-143